pkg/machine/compression/decompress.go

```diff
@@ -1,313 +1,113 @@
 package compression
 
 import (
-	"archive/zip"
-	"bufio"
-	"compress/gzip"
 	"errors"
-	"fmt"
 	"io"
 	"os"
-	"os/exec"
 	"path/filepath"
 	"runtime"
 	"strings"
 
-	"github.com/containers/image/v5/pkg/compression"
 	"github.com/containers/podman/v5/pkg/machine/define"
 	"github.com/containers/podman/v5/utils"
 	"github.com/containers/storage/pkg/archive"
-	crcOs "github.com/crc-org/crc/v2/pkg/os"
-	"github.com/klauspost/compress/zstd"
 	"github.com/sirupsen/logrus"
-	"github.com/ulikunitz/xz"
 )
 
-// Decompress is a generic wrapper for various decompression algos
-// TODO this needs some love.  in the various decompression functions that are
-// called, the same uncompressed path is being opened multiple times.
-func Decompress(localPath *define.VMFile, uncompressedPath string) error {
-	var isZip bool
-	uncompressedFileWriter, err := os.OpenFile(uncompressedPath, os.O_CREATE|os.O_RDWR, 0600)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := uncompressedFileWriter.Close(); err != nil && !errors.Is(err, os.ErrClosed) {
-			logrus.Warnf("unable to close decompressed file %s: %q", uncompressedPath, err)
-		}
-	}()
-	sourceFile, err := localPath.Read()
-	if err != nil {
-		return err
-	}
-	if strings.HasSuffix(localPath.GetPath(), ".zip") {
-		isZip = true
-	}
-	compressionType := archive.DetectCompression(sourceFile)
-
-	prefix := "Extracting compressed file"
-	prefix += ": " + filepath.Base(uncompressedPath)
-	switch compressionType {
-	case archive.Xz:
-		return decompressXZ(prefix, localPath.GetPath(), uncompressedFileWriter)
-	case archive.Uncompressed:
-		if isZip && runtime.GOOS == "windows" {
-			return decompressZip(prefix, localPath.GetPath(), uncompressedFileWriter)
-		}
-		// here we should just do a copy
-		dstFile, err := os.Open(localPath.GetPath())
-		if err != nil {
-			return err
-		}
-		// darwin really struggles with sparse files. being diligent here
-		fmt.Printf("Copying uncompressed file %q to %q/n", localPath.GetPath(), dstFile.Name())
-
-		// Keeping CRC implementation for now, but ideally this could be pruned and
-		// sparsewriter could be used.  in that case, this area needs rework or
-		// sparsewriter be made to honor the *file interface
-		_, err = crcOs.CopySparse(uncompressedFileWriter, dstFile)
-		return err
-	case archive.Gzip:
-		if runtime.GOOS == "darwin" {
-			return decompressGzWithSparse(prefix, localPath, uncompressedFileWriter)
-		}
-		fallthrough
```

Comment:
For the record:

This code from #21768 works right now, but if Zstd ever grew a non-Darwin implementation, Gzip would start failing on Darwin. To me, that kind of a trap is at least worth a comment.

That’s fixed in this PR — but if this PR were abandoned for any reason, I would want to add that comment.